codePost_lib.py

```diff
@@ -0,0 +1,695 @@
+                submission_info=submission,
+                newest_files=files,
+                mode=mode
+            )
+
+    return None
+
+
+def diffscan_submission(api_key, submission_info, newest_files, mode=DEFAULT_MODE):
+
+    # Retrieve a submission's existing files
+    existing_files = {
+        file["name"]: file
+        for file in [
+            get_file(api_key=api_key, file_id=file_id)
+            for file_id in submission_info['files']
+        ]
+    }
+
+    #
+    submission_was_modified = False
+
+    for file in newest_files:
+
+        # Check if file matches existing ones (by matching name and extension)
+        if file["name"] in existing_files and existing_files[file["name"]]["extension"] == file["extension"]:
+
+            if mode["updateExistingFiles"]:
+
+                # FIXME: use hashing/robust method of comparing files
+
+                # Ignore newlines when comparing files, to avoid a trailing newline
+                # registering as a difference
+
+                data_existing = existing_files[file["name"]]["code"].replace(
+                    "\n", "")
+                data_new = file["code"].replace("\n", "")
+
+                if data_existing != data_new:
+
+                    submission_was_modified = True
+                    _print_info(
+                        "Replacing contents of {} (note: all comments will be deleted)")
+
+                    delete_file(api_key=api_key,
+                                file_id=existing_files[file["name"]]["id"])
+                    post_file(
+                        api_key=api_key,
+                        submission_id=submission_info["id"],
+                        filename=file["name"],
+                        content=file["code"],
+                        extension=file["extension"]
+                    )
+
+        else:
+            submission_was_modified = True
+            _print_info("Adding file {}.".format(file["name"]))
+            post_file(
+                api_key=api_key,
+                submission_id=submission_info["id"],
+                filename=file["name"],
+                content=file["code"],
+                extension=file["extension"]
+            )
+
+    if not submission_was_modified:
+        _print_info("Nothing to add or update, submission was left unchanged.")
+
+    return submission_was_modified
+
+###########################################################################################
+
+
+def get_available_courses(api_key, course_name=None, course_period=None):
+    """
+    Returns a list of the available courses/terms to which the user, associated with
+    the provided API key, has access to. Optionally, restrict the results to a specific
+    course and/or period.
+    """
+    auth_headers = {"Authorization": "Token {}".format(api_key)}
+
+    result = None
+
+    try:
+        r = _requests.get(
+            "{}/users/me".format(BASE_URL),
+            headers=auth_headers
```

Comment:
We can replace this call with a new endpoint: `GET /courses/`, which now returns a list of all courses to which the caller has access. I'll document this new endpoint shortly.

The reason for removing `/users/me` is that this endpoint isn't intended to be publicly facing.

Reply:
Sounds good, I'll update that method.